draft-ietf-tls-tls13.md

```diff
@@ -4696,6 +4642,195 @@ a context in all exporter computations, though the value could be empty.
+
+In normal operation, clients will not know which, if any, of these
+mechanisms servers actually implement and therefore MUST only send
+early data which they are willing to have subject to the attacks
+described in {{replay-0rtt}}.
+
+
+## Single-Use Tickets
+
+The simplest form of anti-replay defense is for the server to only
+allow each session ticket once. In order to implement this, the server
+maintains a database of all outstanding valid tickets; deleting each
+ticket from the database as it is used. If an unknown ticket is
+provided, the server falls back to a full handshake as normal.
+
+If the tickets are not self-contained but rather are database keys,
+and these PSKs are deleted upon use, then connections established
+using one PSK enjoy forward security. This is a security advantage for
+all 0-RTT data and for PSK usage when PSK is used without DH.
+
+Because this mechanism requires sharing the session database between
+server nodes in environments with multiple distributed servers,
+in such cases it may be hard to achieve high rates of PSK 0-RTT
+success when compared with self-encrypted tickets which do not
+require consistent server-side storage for PSK-based session
+establishment but do require it for anti-replay if 0-RTT is allowed,
+as described below.
+
+
+## Client Hello Recording
+
+An alternative form of anti-replay is to record a unique value derived
+from the ClientHello (generally either the random value or the PSK
+binder) and reject duplicates. Recording all ClientHellos causes state
+to grow without bound. A server can instead record ClientHellos within
+a given time window and use the "obfuscated_ticket_age" to ensure that
+tickets aren't reused outside that window.
+
+In order to implement this mechanism, a server needs to store the time
+that the server generated the session ticket, offset by an estimate of
+the round trip time between client and server. I.e.,
+
+~~~~
+    adjusted_creation_time = creation time + estimated RTT
+~~~~
+
+This value can be encoded in the ticket, thus avoiding the need to
+keep state for each outstanding ticket. The server can determine the
+client's view of the age of the ticket by subtracting the ticket's
+"ticket_age_add value" from the "obfuscated_ticket_age" parameter in
+the client's "pre_shared_key" extension. The server can determine the
+"expected arrival time" of the ClientHello as:
+
+~~~~
+    expected_arrival_time = adjusted_creation_time + client's ticket age
+~~~~
+
+For a given Client Hello recording window, the server implements anti-replay as
+follows.
+
+1. Verify the PSK binder as described in {{pre-shared-key-extension}}.
+
+2. If the expected_arrival_time is outside the window or the ClientHello
+   matches a known ClientHello then accept the PSK but
+   reject 0-RTT.
+
+3. If the ClientHello matches a known ClientHello then
+   either abort the handshake with an "illegal_parameter" alert
+   or accept the PSK but reject 0-RTT.
```

Comment:
Step 2 says to reject 0-RTT when the ClientHello is known, which is it?  I think that you want to remove that clause from step 2.

Aside: given the plans for NSS, I will be rejecting 0-RTT, but not sending an illegal_parameter alert.  This is because I'm not planning to store the ClientHello and with a false positive chance, we don't want to fail connections like that.  Yes, this is less good, but that's what you get.  (Oh, I see that you have that below, good.)